### PR TITLE
feat: devex improvements to noirjs

### PR DESCRIPTION
**Pull request by signorecello**
# Description

Working on autogenerated docs at #3035 I came up with some improvements to UX and Devex for NoirJS.

## Problem\*

Not really a problem, just addressing some DevEx concerns and UX. Before this PR, standard usage was:

```js
const backend = new BarretenbergBackend(circuit, 4);
const noir = new Noir(circuit, backend);
```

This could be improved

## Summary\*

- It adds the destroy method to proxy to backend_barretenberg, saving user's resources.
- Adds an "options" object that is passed to the initialisation of barretenberg:
```js
// previously
new Backend(circuit, 4) // what's the 4?

// now
new Backend(circuit { numOfThreads : 4}) // oh it's the number of threads!
```
- Adds the `circuit` parameter to the new `execute` function instead of the Noir class - this avoids creating a new Noir class everytime we want to execute something, plus removes the need to always pass a circuit twice (to backend, and to Noir)
- ~~Makes it not export acvm, abi, generateWitness, and others from the index file of noir_js, as a precursor to #3035~~ It will still export these and we'll figure it out with the docs autogenerator
- Separated the Backend type into two interfaces: Backend and BackendInternal so it's clearer what's supposed to be called by the Noir class and what's not yet implemented there. Also precursor to #3035

### CODE REVIEW

**Comment by TomAFrench** [comment]
> Noir only instantiates the Backend once: at init. This makes the user UX faster as it's not called every time we prove or verify something.

We would only instantiate if the backend hadn't already been instantiated previously. This change saves us checking whether a variable is defined or not but means we give up the option of lazy initialization. I don't think that this is a good trade.

**Comment by TomAFrench** [comment]
> Noir doesn't take in the circuit in the constructor. Instead, since we're passing it to the Backend, we can just call this.backend.circuit.

This requires us to always have a backend when interacting with the `Noir` class. I can see situations where you would want to create an interact with a Noir program without instantiating a backend at all (e.g. I want to execute a program to get its return value but I don't want to create a proof), it would be good to be able to just omit the backend in these cases but adding this reliance would prevent this.

**Comment by TomAFrench** [comment]
> Makes it not export acvm, abi, generateWitness, and others from the index file of noir_js

I also think this is not ideal. We should give users the tools to break out of the standard workflow if necessary (we've already run into a couple of cases where this is necessary).

**Comment by TomAFrench** [comment]
> he can always import the acvm and abi packages.

This kinda defeats the point of noir.js, no? If we're expecting users to manage versions of the underlying packages to keep them in sync with each other, and as well as that with the versions included in noir.js then we've just made the problem worse.

**Comment by signorecello** [comment]
I actually raise the same point, but with a different argumentation... If `noir_js` exists just to export correct versions of `acvm` and `abi`, then _that_ would be defeating the purpose. 

In other words, `noir_js` exists exactly for the vast majority of users who wouldn't have a use for importing and managing `acvm` or `abi` versions. It's meant to tackle 90% of the use cases, while not leaving behind advanced users.

We're not excluding advanced developers from using our tools, we're including everybody else.

**Comment by TomAFrench** [comment]
> If noir_js exists just to export correct versions of acvm and abi, then that would be defeating the purpose.

I didn't say that it's the only purpose but a significant motivator for `noir_js` was to provide a package with a known-consistent set of dependencies.

**Comment by signorecello** [comment]
cc #2910

**Comment by signorecello** [comment]
So talking to @kevaundray we decided to export everything (and shut up the docs autogenerator) since we're not sure which methods will be used elsewhere.

Once that's cleared up, we can narrow them as part of #3078 

Let me just make some changes and I'll push & resolve conflicts. Thanks @TomAFrench for the discussion

**Comment by Savio-Sou** [comment]
Thanks @signorecello! Do request for re-reviews when ready.

File: tooling/noir_js/src/index.ts [hunk 1]
```diff
@@ -1,8 +1,59 @@
...
+    // If these are available, then we are in the
+    // web environment. For the node environment, this
+    // is a no-op.
+    if (typeof initAbi === 'function') {
+      await Promise.all([initAbi(), initACVM()]);
+    }
+  }
+
+  async destroy(): Promise<void> {
+    if (!this.backend) throw new Error('No backend to destroy');
```
Comment (TomAFrench) [inline tooling/noir_js/src/index.ts:23]: This should just be a noop in this case rather than an error.

File: tooling/noir_js/src/index.ts [hunk 2]
```diff
@@ -1,8 +1,59 @@
...
+
+  private getBackend(): Backend {
+    if (this.backend === undefined) throw new Error('Operation requires a backend but none was provided');
+    return this.backend;
+  }
+
+  // Initial inputs to your program
+  async execute(
+    inputs: abi.InputMap,
+    circuit?: CompiledCircuit,
```
Comment (TomAFrench) [inline tooling/noir_js/src/index.ts:36]: I think it's a better API to instantiate the `Noir` class with a `CompiledCircuit` and use that in this function rather than having to pass in a circuit each time.

This is inconsistent when `Noir` "knows" about the circuit you're interacting with when you're proving/verifying but not for execution.

File: tooling/noir_js/test/node/cjs.test.cjs [hunk 3]
```diff
@@ -21,8 +23,8 @@ it('string input and number input are the same', async () => {
     x: 2,
     y: 3,
   };
-  const solvedWitnessString = await new Noir(assert_lt_json).execute(inputsString);
-  const solvedWitnessNumber = await new Noir(assert_lt_json).execute(inputsNumber);
+  const solvedWitnessString = await new Noir(assert_lt_json).execute(inputsString, assert_lt_program);
+  const solvedWitnessNumber = await new Noir(assert_lt_json).execute(inputsNumber, assert_lt_program);
```
Comment (TomAFrench) [inline tooling/noir_js/test/node/cjs.test.cjs:27]: These are broken. You're passing the circuit in twice.

File: tooling/noir_js_backend_barretenberg/src/index.ts [hunk 4]
```diff
@@ -1,25 +1,25 @@
...
 
 // This is the number of bytes in a UltraPlonk proof
 // minus the public inputs.
 const numBytesInProofWithoutPublicInputs: number = 2144;
 
 export class BarretenbergBackend implements Backend {
-  // These type assertions are used so that we don't
-  // have to initialize `api` and `acirComposer` in the constructor.
-  // These are initialized asynchronously in the `init` function,
-  // constructors cannot be asynchronous which is why we do this.
```
Comment (TomAFrench) [inline tooling/noir_js_backend_barretenberg/src/index.ts:13]: Why is this removed?

**Review by TomAFrench — COMMENTED** [review]
I think this PR is muddling what the `Noir` class represents, it encapsulates a Noir program and should expose methods to interact with this program. `new Noir()` then doesn't really have any meaning and any method calls on `Noir()` could just be regular functions.

I think making the `Backend` be the source of truth on the circuit in question is also a misstep. The `Backend` only needs the circuit to determine how much of the CRS to load and is otherwise generic, ideally I should be able to reuse the same backend object for creating proofs for two different circuits but this PR takes us further away from this.

File: tooling/noir_js/test/node/e2e.test.ts [hunk 5]
```diff
@@ -6,148 +6,118 @@ import { CompiledCircuit } from '@noir-lang/types';
...
-// `RuntimeError: null function or function signature mismatch`
-//
-// This happens when we we create a proof with one barretenberg instance and
-// try to verify it with another.
-//
-// If this bug is fixed, we can remove this test and split barretenberg into
-// a prover and verifier class to more accurately reflect what happens in production.
-//
-// If its not fixable, we can leave it in as documentation of this behavior.
-it('[BUG] -- bb.js null function or function signature mismatch (different instance) ', async () => {
```
Comment (TomAFrench) [inline tooling/noir_js/test/node/e2e.test.ts:84]: Can you explain what's going on with these deletions? I'm not clear on how these connect with the rest of the PR.

File: tooling/noir_js_backend_barretenberg/src/index.ts (hunk 4 above)
Comment (kobyhallx) [reply tooling/noir_js_backend_barretenberg/src/index.ts:13]: Those seemd to be in wrong line

**Comment by kobyhallx** [comment]
I have similar observations on the surface to Tom. Seems that we do not need `noir_js` which by design should represent noir program which can be processed with initialised backend. With proposed changes it feels that we would rather benefit from backend package that would be used by developer while taking artefacts from `noir` (or hide them within it's API)?

**Comment by signorecello** [comment]
So I've talked with @TomAFrench and it seems like we could drop this in favour of lazy initialization of the CRS which would be a more dynamic way of achieving the same (not passing the circuit to both instances).

I'll be spinning off of this to add the destroy method and the options object in another PR.

Thanks @kobyhallx for the review

File: tooling/noir_js/test/node/e2e.test.ts (hunk 5 above)
Comment (signorecello) [reply tooling/noir_js/test/node/e2e.test.ts:84]: yeah github is fucking up, file is fine

File: tooling/noir_js_backend_barretenberg/src/index.ts (hunk 4 above)
Comment (signorecello) [reply tooling/noir_js_backend_barretenberg/src/index.ts:13]: yep somehow it says I've deleted wrong lines but if you see the code, they're actually there
